### PR TITLE
Updated rms to factor in weighting

### LINKED ISSUE #814: simple rms.rmsd() ignores weights option
The `analysis.rms.rmsd` function doesn't weight a calculated rmsd when the weights are given

### Currently version of MDAnalysis:
(run `python -c "import MDAnalysis as mda; print(mda.__version__)"`)
0.15dev

### PR DESCRIPTION
**Pull request by jdetle**
Fixes #814 

Changes made in this Pull Request:
- Changed to return a weighted RMSD

I'll upload some tests that I'm working on tomorrow morning. Right now this doesn't seem to be doing what I would like.

PR Checklist
------------
 - [x] Tests?
 - [ ] Docs?
 - [x] CHANGELOG updated?
 - [x] Issue raised/referenced?

### CODE REVIEW

**Comment by jdetle** [comment]
For example, I would expect a weight array of [1, 0, ..., 0_n] to yield the RMSD between both atoms in the first coordinate, but right now that is not the case. I imagine this has to do with the order of operations, and where N is factored in.

**Comment by kain88-de** [comment]
what do you get when you enable superposition with those weights?

**Comment by jdetle** [comment]
An error, darn. Looks like the np.mean has to stay to the right superposition.

**Comment by jdetle** [comment]
I should probably test before I commit -.-

**Comment by jdetle** [comment]
Referencing #822. I don't know if this is a bug or not, but I found this while trying to write tests for weighting. If we want to use a single atom from a trajectory frame for to calculate an rmsd, the shape changes from being (#number_atoms, 3) to being simply (3,). This means that instead of dividing by 1, as RMSD should on a 1 atom calculation, it divides by 3 (given that N = b.shape[0]). Do we want RMSD to work for single atoms or is it an irrelevant edge case?

**Comment by jdetle** [comment]
If someone wants to check this I think it is ready to be merged.

File: package/MDAnalysis/analysis/rms.py [hunk 1]
```diff
@@ -204,7 +204,13 @@ def rmsd(a, b, weights=None, center=False, superposition=False):
         return qcp.CalcRMSDRotationalMatrix(a.T, b.T, N, None,
                                             relative_weights)
     else:
-        return np.sqrt(np.sum((a - b) ** 2) / N)
+        if weights is not None:
+            #return np.sqrt(np.sum(relative_weights[:,np.newaxis]
+            #        * ( ( a - b ) ** 2) ) / N)
+            return np.sqrt(np.sum(relative_weights[:,np.newaxis]
+                * ( ( a - b ) ** 2) ) / N)
```
Comment (kain88-de) [inline package/MDAnalysis/analysis/rms.py]: no spaces between braces

File: package/MDAnalysis/analysis/rms.py [hunk 2]
```diff
@@ -204,7 +204,13 @@ def rmsd(a, b, weights=None, center=False, superposition=False):
         return qcp.CalcRMSDRotationalMatrix(a.T, b.T, N, None,
                                             relative_weights)
     else:
-        return np.sqrt(np.sum((a - b) ** 2) / N)
+        if weights is not None:
+            #return np.sqrt(np.sum(relative_weights[:,np.newaxis]
+            #        * ( ( a - b ) ** 2) ) / N)
+            return np.sqrt(np.sum(relative_weights[:,np.newaxis]
```
Comment (kain88-de) [inline package/MDAnalysis/analysis/rms.py]: space after a comma

File: package/MDAnalysis/analysis/rms.py [hunk 3]
```diff
@@ -204,7 +204,13 @@ def rmsd(a, b, weights=None, center=False, superposition=False):
         return qcp.CalcRMSDRotationalMatrix(a.T, b.T, N, None,
                                             relative_weights)
     else:
-        return np.sqrt(np.sum((a - b) ** 2) / N)
+        if weights is not None:
+            #return np.sqrt(np.sum(relative_weights[:,np.newaxis]
```
Comment (kain88-de) [inline package/MDAnalysis/analysis/rms.py]: please remove the comment

File: testsuite/MDAnalysisTests/analysis/test_rms.py [hunk 4]
```diff
@@ -73,6 +73,14 @@ def test_superposition(self):
         rmsd = rms.rmsd(a, b, superposition=True)
         assert_almost_equal(rmsd, 6.820321761927005)
 
+    def test_weights(self):
```
Comment (kain88-de) [inline testsuite/MDAnalysisTests/analysis/test_rms.py]: please add a tests that you get the same results using weights with and without `superposition=True`. Then you are comparing the both RMSD algorithms to each other.

I have shown in the tests above how this is done.

Comment (jdetle) [inline testsuite/MDAnalysisTests/analysis/test_rms.py]: Will do, thanks.

**Comment by jdetle** [comment]
@kain88-de 
Weights with superposition runs into an issue and I don't know if it's an acceptable failure. 
This test
```
def test_weights_and_superposition_2(self):
        weights = np.zeros(len(self.u.trajectory[0]))
        weights[:100] = 1
        weighted = rms.rmsd(self.u.trajectory[0], self.u.trajectory[1],
            weights=weights, superposition=True)
        firstCoords = rms.rmsd(self.u.trajectory[0][:100], self.u.trajectory[1][:100],
            superposition=True)
        assert_almost_equal(weighted, firstCoords)
```
This fails with:
```
Arrays are not almost equal to 7 decimals
 ACTUAL: 9.333334653700027e-08
 DESIRED: 2.697398304697218e-07
```
Is this an acceptable error or something we want to fix?

**Comment by kain88-de** [comment]
The error is ok both are essentially zero. Comparing to zeros is tricky and way to help that is to add 1 to both numbers and do the comparison afterwards (add a comment why you do that). if this isn't enough to make the test pass reduce the required precision.

**Comment by jdetle** [comment]
Alright, we're covered then.

**Comment by kain88-de** [comment]
could you please squash the commit you have marked as tosquash

**Comment by jdetle** [comment]
Oh shoot. Missed that. Just went out but I'll be back to my computer in a
few hours.
On Sun, Apr 17, 2016 at 11:14 AM kain88-de <notifications@github.com> wrote:

> could you please squash the commit you have marked as tosquash
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/MDAnalysis/mdanalysis/pull/826#issuecomment-211074837>
>
-- 
Have a wonderful day,
John J. Detlefs

**Comment by kain88-de** [comment]
don't worry no harm done. You've broke the tracis build anyway :wink:

**Comment by jdetle** [comment]
I have no idea what broke, tests pass on my machine. Is there any way to see where the failure occurs with travis?

**Comment by richardjgowers** [comment]
@jdetle you should be able to click 'show all checks' then 'details' to see the build.

https://travis-ci.org/MDAnalysis/mdanalysis/jobs/123767820

Looks like you're comparing to zero and the precision isn't there? (ie change the test, E-06 is fine)

**Comment by jdetle** [comment]
Thanks @richardjgowers, for some reason that test passed on my machine with 7 decimals but its updated now and should work.

**Comment by kain88-de** [comment]
@jdetle welcome to the wonderful world of floating point arithmetic with all the random rounding errors you could want.

**Pull request merged**